bin/test.sh

```diff
@@ -7,21 +7,30 @@ set -euo pipefail
 STACK="${1}"
 BASE_IMAGE="heroku/${STACK/-/:}-build"
 OUTPUT_IMAGE="redis-test-${STACK}"
+REDIS_CI_TLS=${REDIS_CI_TLS:-}
 
 echo "Building buildpack on stack ${STACK}...with redis version ${REDIS_VERSION}"
 
 docker build \
+    --no-cache \
     --build-arg "BASE_IMAGE=${BASE_IMAGE}" \
     ${REDIS_VERSION:+--build-arg "REDIS_VERSION=${REDIS_VERSION}"} \
+    ${REDIS_CI_TLS:+--build-arg "REDIS_CI_TLS=${REDIS_CI_TLS}"} \
     -t "${OUTPUT_IMAGE}" \
     .
 
 echo "Checking redis-server presence and version..."
 
+if [ -n "$REDIS_CI_TLS" ]; then
+    REDIS_TLS_ARGS="--tls --insecure"
+else
+    REDIS_TLS_ARGS=""
+fi
+
 # Redis <4 does not support connection URLs so REDIS_URL has to be parsed:
 # https://stackoverflow.com/questions/38271281/can-i-use-redis-cli-with-a-connection-url
-REDIS_CONNECTION_ARGS="\$(echo \${REDIS_URL} | sed 's_redis://\(.*\):\(.*\)@\(.*\):\(.*\)/_-h \3 -p \4 -a \2_')"
-TEST_COMMAND="source .profile.d/redis.sh && sleep 1 && redis-cli ${REDIS_CONNECTION_ARGS} info | grep redis_version:${REDIS_VERSION:-}"
+REDIS_CONNECTION_ARGS="\$(echo \${REDIS_URL} | sed 's_rediss\{0,1\}://\(.*\):\(.*\)@\(.*\):\(.*\)_-h \3 -p \4_')"
+TEST_COMMAND="source .profile.d/redis.sh && sleep 1 && redis-cli ${REDIS_CONNECTION_ARGS} ${REDIS_TLS_ARGS} info | grep redis_version:${REDIS_VERSION:-}"
```

Comment:
Why not use `-u` now we only support the right versions?

In testing, `redis-cli` doesn't seem to do the right thing when presented with the URI:

```
root@7eb864a51ceb:/app# redis-cli --tls --insecure -u "$REDIS_URL"
Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe.
AUTH failed: WRONGPASS invalid username-password pair or user is disabled.
```

Removing the leading colon fixes this (at least on 7.0.11) which indicates the `user` field on the `cliConnInfo` struct is not `NULL` when parsed like this.

It doesn't make a huge difference - most libraries parse this stuff out anyway. Could fix up with a:

```
REDIS_URL=$(echo "$REDIS_URL" | sed 's_/:_/_g')
```

but that gets weird across major versions.

Reply:
This actually bothered me so much I created a small repro: https://gist.github.com/mble/505225ff16fe3ce30a5f372a95057081

Reply:
This is probably a very small bug in Redis (there isn't a lot of surface area testing `redis-cli`). The correct change is probably something like:

```patch
diff --git a/src/cli_common.c b/src/cli_common.c
index 421e7d34a..08d43d276 100644
--- a/src/cli_common.c
+++ b/src/cli_common.c
@@ -339,8 +339,10 @@ void parseRedisUri(const char *uri, const char* tool_name, cliConnInfo *connInfo
     /* Extract user info. */
     if ((userinfo = strchr(curr,'@'))) {
         if ((username = strchr(curr, ':')) && username < userinfo) {
-            connInfo->user = percentDecode(curr, username - curr);
-            curr = username + 1;
+            if (username != curr) {
+                connInfo->user = percentDecode(curr, username - curr);
+                curr = username + 1;
+            }
         }
 
         connInfo->auth = percentDecode(curr, userinfo - curr);

```